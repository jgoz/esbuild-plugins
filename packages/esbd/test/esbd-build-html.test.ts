import { node } from 'execa';
import fs from 'fs';
import path from 'path';
import { beforeAll, describe, expect, it } from 'vitest';

import type { EsbdConfig } from '../lib';

import type { BuildWithHTMLOutput } from './config/serializer';

interface BuildWithHTMLOptions {
  config: Partial<EsbdConfig>;
  files: { [file: string]: string };
}

const TEST_ROOT = path.join(__dirname, '..', 'test-results', 'html');

async function buildWithHTML(options: BuildWithHTMLOptions): Promise<BuildWithHTMLOutput> {
  const absWorkingDir = await fs.promises.mkdtemp(path.join(TEST_ROOT, 'test-'));
  const absOutDir = path.join(absWorkingDir, 'out');

  await fs.promises.mkdir(absOutDir, { recursive: true });

  const index = Object.keys(options.files).find(file => file.endsWith('index.html'));
  if (!index) {
    throw new Error('index.html is required');
  }

  const config: EsbdConfig = {
    format: 'esm',
    metafile: true,
    splitting: true,
    sourcemap: false,
    logLevel: 'warning',
    ...options.config,

    absWorkingDir,
    outdir: './out',
    entryPoints: {
      'index.html': index,
    },
  };

  const bundleFile = path.join(absWorkingDir, 'bundle.js');
  const writeBundle = fs.promises.writeFile(
    bundleFile,
    `require('../../../lib').configure(${JSON.stringify(config)});`,
  );

  const writeFiles = Object.entries(options.files).map(async ([file, content]) => {
    const absFilePath = path.join(absWorkingDir, file);
    await fs.promises.mkdir(path.dirname(absFilePath), { recursive: true });
    await fs.promises.writeFile(absFilePath, content, { encoding: 'utf-8' });
  });

  await Promise.all([...writeFiles, writeBundle]);

  const proc = node(bundleFile, ['build'], {
    encoding: 'utf8',
    reject: false,
    cwd: absWorkingDir,
    env: { ...process.env, NO_COLOR: '1' },
  });

  const { stderr, stdout } = await proc;

  return { cwd: absWorkingDir, outdir: absOutDir, stdout, stderr };
}

describe('build command (html entry)', () => {
  beforeAll(async () => {
    await fs.promises.mkdir(TEST_ROOT, { recursive: true });
    return async () => {
      await fs.promises.rm(TEST_ROOT, { recursive: true });
    };
  });

  it('builds a simple HTML entry point', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
        },
        files: {
          'src/index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <script defer type="module" src="./entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from HTML', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
        },
        files: {
          'src/index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <link rel="stylesheet" href="../styles/entry.css" />
                <script defer type="module" src="./entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'styles/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'styles/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from JS', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
        },
        files: {
          'index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <script defer type="module" src="./src/entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'src/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'src/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            import './entry.css';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced assets from HTML', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
        },
        files: {
          'index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <link rel="apple-touch-icon" href="./assets/favicon.png"/>
                <script defer type="module" src="./src/entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'assets/favicon.png': 'IMA FAVICON',
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            function App() { return <div>Hello world</div>; }
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced assets from style tags', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
        },
        files: {
          'src/index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <style>
                  body {
                    background: url(../assets/cats.jpg);
                  }
                </style>
                <script defer type="module" src="./entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'assets/cats.jpg': 'MEOW',
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            function App() { return <div>Hello world</div>; }
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('supports automatic react runtime', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
          jsx: 'automatic',
        },
        files: {
          'src/index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <script defer type="module" src="./entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('writes integrity hashes if requested', () => {
    return expect(
      buildWithHTML({
        config: {
          external: ['react', 'react-dom'],
          integrity: 'sha256',
        },
        files: {
          'src/index.html': `
            <!DOCTYPE html>
            <html>
              <head>
                <style>
                  body {
                    background: url(../assets/cats.jpg);
                  }
                </style>
                <link rel="apple-touch-icon" href="../assets/favicon.png"/>
                <link rel="stylesheet" href="../styles/entry.css" />
                <script defer type="module" src="./entry.tsx"></script>
              </head>
              <body><div id='root'></div></body>
            </html>
          `,
          'assets/cats.jpg': 'MEOW',
          'assets/favicon.png': 'IMA FAVICON',
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            function App() { return <div>Hello world</div>; }
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'styles/entry.css': `
            body { background: red; }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });
});
