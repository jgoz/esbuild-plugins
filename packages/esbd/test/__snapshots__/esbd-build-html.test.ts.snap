// Vitest Snapshot v1

exports[`build command (html entry) > builds a simple HTML entry point 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <script defer="" type="module" src="entry.js"></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > includes referenced CSS from HTML 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.0.css
---------------------------------
/* styles/app.css */
.app {
  background: green;
}

/* styles/entry.css */
body {
  background: red;
}

---------------------------------
entry.1.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link rel="stylesheet" href="entry.0.css" />
    <script defer="" type="module" src="entry.1.js"></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > includes referenced CSS from JS 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.css
---------------------------------
/* src/app.css */
.app {
  background: green;
}

/* src/entry.css */
body {
  background: red;
}

---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <script defer="" type="module" src="entry.js"></script>
    <link href="entry.css" rel="stylesheet" />
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > includes referenced assets from HTML 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
favicon.png
---------------------------------
IMA FAVICON
---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link rel="apple-touch-icon" href="favicon.png" />
    <script defer="" type="module" src="entry.js"></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > includes referenced assets from style tags 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
cats.jpg
---------------------------------
MEOW
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <style>
      body {
        background: url(cats.jpg);
      }
    </style>
    <script defer="" type="module" src="entry.js"></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > supports automatic react runtime 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
import { jsx } from "react/jsx-runtime";
function App() {
  return /* @__PURE__ */ jsx("div", {
    children: "Hello world",
  });
}

// src/entry.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
ReactDOM.render(/* @__PURE__ */ jsx2(App, {}), document.getElementById("root"));

---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <script defer="" type="module" src="entry.js"></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;

exports[`build command (html entry) > writes integrity hashes if requested 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished index.html with 0 error(s) and 0 warning(s) in XX time
---------------------------------
cats.jpg
---------------------------------
MEOW
---------------------------------
entry.0.css
---------------------------------
/* styles/entry.css */
body {
  background: red;
}

---------------------------------
entry.1.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
favicon.png
---------------------------------
IMA FAVICON
---------------------------------
index.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <style>
      body {
        background: url(cats.jpg);
      }
    </style>
    <link
      rel="apple-touch-icon"
      href="favicon.png"
      integrity="sha256-m/XpYP27F4yCq5KBTKxEzP5x4gZBZjo5zeuDXbm+ypw="
    />
    <link
      rel="stylesheet"
      href="entry.0.css"
      integrity="sha256-h7N0aRQFgi82y9VoYyAN3AlXuyhJ58bQ1uWjjffOWm4="
    />
    <script
      defer=""
      type="module"
      src="entry.1.js"
      integrity="sha256-NvQobgjBDGRnA/TwQI2waRjuU25hnAjteV1NhFBmjbo="
    ></script>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>

`;
